File: backend/app/api/export.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Response
from sqlalchemy.orm import Session

from app.auth.dependencies import AuthenticatedSession, get_authenticated_session
from app.database.session import get_db
from app.services.export import export_csv, export_json


router = APIRouter(prefix="/api/v1/export", tags=["export"])


@router.get("/csv")
def csv_download(
    _authenticated: AuthenticatedSession = Depends(get_authenticated_session),
    database: Session = Depends(get_db),
) -> Response:
    return Response(
        export_csv(database),
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": 'attachment; filename="tapledger-transactions.csv"'},
    )


@router.get("/json")
def json_download(
    _authenticated: AuthenticatedSession = Depends(get_authenticated_session),
    database: Session = Depends(get_db),
) -> Response:
    return Response(
        export_json(database),
        media_type="application/json; charset=utf-8",
        headers={"Content-Disposition": 'attachment; filename="tapledger-transactions.json"'},
    )

File: frontend/src/worker/routes/export.ts
import { Hono } from 'hono'
import { requireSession } from '../middleware/auth'
import { ExportService } from '../services/exportService'
import type { AppEnvironment } from '../types'

function attachment(name: string): Record<string, string> {
  return { 'Content-Disposition': `attachment; filename="${name}"`, 'Cache-Control': 'no-store' }
}

export const exportRoutes = new Hono<AppEnvironment>()
exportRoutes.use('/export/*', requireSession)

exportRoutes.get('/export/csv', async (c) => c.body(await new ExportService(c.env.DB).csv(), 200, {
  ...attachment('tapledger-transactions.csv'),
  'Content-Type': 'text/csv; charset=utf-8',
}))

exportRoutes.get('/export/json', async (c) => c.body(await new ExportService(c.env.DB).json(), 200, {
  ...attachment('tapledger-transactions.json'),
  'Content-Type': 'application/json; charset=utf-8',
}))

exportRoutes.get('/export/backup', async (c) => c.body(JSON.stringify(await new ExportService(c.env.DB).backupDocument(), null, 2), 200, {
  ...attachment('tapledger-backup.json'),
  'Content-Type': 'application/json; charset=utf-8',
}))
